library/alloc/src/boxed.rs

```diff
@@ -472,13 +720,30 @@ impl<T: ?Sized> Box<T> {
     )]
     #[inline]
     #[doc(hidden)]
-    pub fn into_unique(b: Box<T>) -> Unique<T> {
+    pub fn into_unique(b: Self) -> (Unique<T>, A) {
         // Box is recognized as a "unique pointer" by Stacked Borrows, but internally it is a
         // raw pointer for the type system. Turning it directly into a raw pointer would not be
         // recognized as "releasing" the unique pointer to permit aliased raw accesses,
-        // so all raw pointer methods go through `leak` which creates a (unique)
-        // mutable reference. Turning *that* to a raw pointer behaves correctly.
-        Box::leak(b).into()
+        // so all raw pointer methods have to leak the box. Turning *that* to a raw pointer
+        // behaves correctly.
+        let b = mem::ManuallyDrop::new(b);
+
+        // The box is unitiliazed later when moving out the allocator. The pointer is stored
+        // beforehand.
+        let ptr = b.0;
+        let alloc = unsafe { ptr::read(&b.1) };
+        (ptr, alloc)
+    }
+
+    /// Returns a reference to the underlying allocator.
+    ///
+    /// Note: this is an associated function, which means that you have
+    /// to call it as `Box::alloc(&b)` instead of `b.alloc()`. This
+    /// is so that there is no conflict with a method on the inner type.
+    #[unstable(feature = "allocator_api", issue = "32838")]
+    #[inline]
+    pub fn alloc(b: &Self) -> &A {
```

Comment:
This function seems poorly named to me since `alloc` sounds like a verb (`Box::alloc` sounds like it should allocate memory). Maybe rename it to `alloc_ref` or `allocator` to make it clearer.